provider_client.go

```diff
@@ -215,13 +200,47 @@ func (client *ProviderClient) Request(method, url string, options RequestOpts) (
 	if !ok {
 		body, _ := ioutil.ReadAll(resp.Body)
 		resp.Body.Close()
-		return resp, &UnexpectedResponseCodeError{
+		respErr := &UnexpectedResponseCodeError{
 			URL:      url,
 			Method:   method,
 			Expected: options.OkCodes,
 			Actual:   resp.StatusCode,
 			Body:     body,
 		}
+
+		errType := options.ErrorType
+		switch resp.StatusCode {
+		case http.StatusUnauthorized:
+			if client.ReauthFunc != nil {
+				err = client.ReauthFunc()
+				if err != nil {
+					return nil, fmt.Errorf("Error trying to re-authenticate: %s", err)
+				}
+				if options.RawBody != nil {
+					options.RawBody.Seek(0, 0)
+				}
+				resp, err = client.Request(method, url, options)
+				if err != nil {
+					return nil, fmt.Errorf("Successfully re-authenticated, but got error executing request: %s", err)
+				}
+			}
+			err = defaultError401{}
+			if _, ok := errType.(Error401er); ok {
+				err = errType.(Error401er).Error401(respErr)
+			}
+		case http.StatusNotFound:
+			err = defaultError404{}
+			if _, ok := errType.(Error404er); ok {
+				err = errType.(Error404er).Error404(respErr)
+			}
+		case http.StatusMethodNotAllowed:
+			err = defaultError405{}
+			if _, ok := errType.(Error405er); ok {
+				err = errType.(Error405er).Error405(respErr)
+			}
```

Comment:
:+1: Not bad, not bad.

I think you could save yourself the double cast like so:

```go
if err405, ok := errType.(Error405er); ok {
  err = err405.Error405(respErr)
}
```